python/TestHarness/tests/test_QueuePBS.py

```diff
@@ -0,0 +1,418 @@
+                        raise Exception('Failed: QueueManager launched one job, but received several results')
+
+            # Secondary checks; lets verify the session file is saving
+            # the results properly
+            if os.path.exists(self.pbs_session_file):
+                with open(self.pbs_session_file, 'r') as session_file:
+                    session = json.load(session_file)
+            else:
+                raise Exception('Failed: Session File was not created')
+
+            for key, value in session.iteritems():
+                if key != 'QUEUEMANAGER':
+                    self.assertRegexpMatches(session[key]['caveat_message'], status)
+
+    def testSessionFile(self):
+        """
+        Verify the session file is what we expect
+        """
+
+        # Launch jobs to create a session file for us to use
+        self.testPBSGoodLaunch()
+
+        if os.path.exists(self.pbs_session_file):
+            with open(self.pbs_session_file, 'r') as session_file:
+                session = json.load(session_file)
+        else:
+            raise Exception('Failed: Session File was not created')
+
+        # Iterate over the session file and do some checks
+        if 'QUEUEMANAGER' not in session.keys():
+            raise Exception('Failed: QUEUEMANAGER configuration key not found')
+
+        checks = ['job_id',
+                  'queue_script',
+                  'job_name',
+                  'working_dir',
+                  'caveat_message',
+                  'status_bucket']
+
+        for key, value in session.iteritems():
+            if key != 'QUEUEMANAGER':
+                if set(checks) - set(value.keys()):
+                    raise Exception('Failed: Invalid items present in session or this unittest is out of date!')
+
+    def testCopyFiles(self):
+        """
+        Test the QueueManagers ability to copy specific files when requested.
+
+        We will create two testers with a single dependency. The first tester,
+        when launched, should not have copied any extra files, and that will
+        be our first test.
+
+        The second test does copy extra files.
+        """
+
+        with self.harnessDAG() as harness_dag:
+            (harness, job_dag) = harness_dag
+
+            for job_id, job in self.yieldJobs(job_dag):
+
+                self._mockSinglePBSLaunch(harness, job, '%d.nowhere.org' % (job_id))
+
+                # we should not have extra files copied after queueing the first job
+                if job_id == 0 and os.path.exists(os.path.join(self.session_dir, 'queue_job_copy_file_test')):
+                    raise Exception('Failed: A file was copied that should not have been')
+
+                # but our second job should have copied some extra stuff over
+                elif job_id == 1 and not os.path.exists(os.path.join(self.session_dir, 'queue_job_copy_file_test')):
+                    raise Exception('Failed: A file was not copied that should have been')
```

Comment:
Should probably use what ever file is listed in the test file instead of arbitrarily entering it here. And perhaps test multiple files.

```python
    for file in specs['copy_files']:
```